GH-1764: Add docs to the core modules.

mpapierski opened the pull request:
Ref #1764 

This PR adds rust docs to the `core` module and its children.

Commit:
Document core module of the EE crate.

Commit:
Update test schema json file.

Commit:
Merge branch 'dev' into gh-1764-core-docs-1

 Conflicts:
	execution_engine/src/core/engine_state/engine_config.rs
	execution_engine/src/core/engine_state/genesis.rs
	execution_engine/src/core/engine_state/mod.rs
	execution_engine/src/core/engine_state/upgrade.rs
	execution_engine/src/core/runtime/mod.rs
	execution_engine/src/core/runtime_context/mod.rs
	execution_engine/src/lib.rs
	execution_engine/src/storage.rs
	node/src/components/rpc_server/rpcs/state.rs

Commit:
Remove reference of ProtocolData.

goral09 commented on execution_engine/CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ All notable changes to this project will be documented in this file.  The format
 * Support building and testing using stable Rust.
 * Increase price of `create_purse` to 2.5CSPR.
 * Improve doc comments to clarify behavior of the bidding functionality.
+* Documented `core` module and children.
```
Is it all `core` modules or only the EE one?

Fraser999 replied:
Only core from EE, but this changelog is specific to EE, so should be fine as is I think

dwerner commented on execution_engine/src/core/engine_state/executable_deploy_item.rs:
```diff
@@ -1,3 +1,4 @@
+//! Ways a smart contract execution can be expressed as.
```
```suggestion
//! Smart contract execution parameters.
```

dwerner commented on execution_engine/src/core/engine_state/executable_deploy_item.rs:
```diff
@@ -752,31 +789,52 @@ impl Distribution<ExecutableDeployItem> for Standard {
     pub account_hash: AccountHash,
+    /// Key pointing to the entity we will be running as.
     pub base_key: Key,
+    /// An instance of WASM module.
```
```suggestion
    /// An instance of the WASM module.
```

dwerner commented on execution_engine/src/core/engine_state/execute_request.rs:
```diff
@@ -1,20 +1,28 @@
+//! Code supporting an contract execution request.
```
```suggestion
//! Code supporting a contract execution request.
```

dwerner commented on execution_engine/src/core/engine_state/execute_request.rs:
```diff
@@ -1,20 +1,28 @@
 use super::deploy_item::DeployItem;
 use crate::shared::newtypes::Blake2bHash;
 
+/// Represents a contract execution request that contains multiple deploys.
```
```suggestion
/// Represents a contract execution request that can contain multiple deploys.
```

dwerner commented on execution_engine/src/core/engine_state/execution_result.rs:
```diff
@@ -43,19 +44,28 @@ fn make_payment_error_effects(
 pub enum ExecutionResult {
     /// An error condition that happened during execution
     Failure {
+        /// Error condition.
```
```suggestion
        /// Error causing this `Failure` variant.
```

dwerner commented on execution_engine/src/core/engine_state/execution_result.rs:
```diff
@@ -73,6 +83,7 @@ impl Default for ExecutionResult {
 /// A type alias that represents multiple execution results.
 pub type ExecutionResults = VecDeque<ExecutionResult>;
 
+/// Indicates the forced payme
```
Incomplete?

mpapierski replied:
Addressed

dwerner commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -1088,6 +1140,14 @@ where
 
+    /// Executes a deploy.
+    ///
+    /// A deploy execution consists of running a payment code which is expected to deposit funds
```
```suggestion
    /// A deploy execution consists of running payment code which is expected to deposit funds
```

dwerner commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -1088,6 +1140,14 @@ where
+    /// Executes a deploy.
+    ///
+    /// A deploy execution consists of running a payment code which is expected to deposit funds
+    /// into a payment purse, and then running a session code with a specific gas limit. For running
```
```suggestion
    /// into a payment purse, and then running session code with a specific gas limit. For running
```

dwerner commented on execution_engine/src/core/engine_state/mod.rs:
```diff
@@ -1678,6 +1746,7 @@ where
         Ok(missing_descendant_trie_keys)
     }
 
+    /// Looks up for a list of missing root hashes.
```
```suggestion
    /// Performs a lookup for a list of missing root hashes.
```

dwerner commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -68,6 +70,15 @@ pub struct Runtime<'a, R> {
+/// This ensures that a memory instance is properly resolved into a pre-allocated memory area, and a
+/// host function resolver is attached to the module.
+///
+/// WASM module is also validated to not have a "start" section as we currently don't support
```
```suggestion
/// The WASM module is also validated to not have a "start" section as we currently don't support
```

dwerner commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -1005,29 +1020,35 @@ where
         &self.call_stack
     }
 
+    /// Returns bytes from WASM memory instance.
```
```suggestion
    /// Returns bytes from the WASM memory instance.
```

dwerner commented:
Other than my requested minor changes, lgtm.

Commit:
Merge branch 'dev' into gh-1764-core-docs-1

 Conflicts:
	execution_engine/CHANGELOG.md
	execution_engine/src/core/engine_state/balance.rs
	execution_engine/src/core/engine_state/query.rs
	execution_engine/src/core/engine_state/step.rs
	execution_engine/src/core/execution/error.rs
	node/src/components/event_stream_server/tests.rs
	node/src/components/rpc_server/rpcs/state.rs
	node/src/reactor/participating/tests.rs

Commit:
Update missing rustdocs after merge.

mpapierski committed:
Apply suggestions from code review

Co-authored-by: Daniel N. Werner <dan@casperlabs.io>

Commit:
Apply rest of feedback from @dwerner

luis-casper committed:
Implementing code comment fixes.

I am also adding a few TODOs that need some attention from people that
are more familiar with the EE.

luis-casper committed:
Fixing some TODOs.

luis-casper committed:
Converting MS-DOS end-of-line to UNIX.

Commit:
Clarify comment on precomputed payment errors.

Commit:
Merge branch 'dev' into gh-1764-core-docs-1

 Conflicts:
	execution_engine/src/core/engine_state/balance.rs
	execution_engine/src/core/engine_state/era_validators.rs
	execution_engine/src/core/engine_state/error.rs
	execution_engine/src/core/engine_state/execute_request.rs
	execution_engine/src/core/engine_state/genesis.rs
	execution_engine/src/core/engine_state/mod.rs
	execution_engine/src/core/engine_state/op.rs
	execution_engine/src/core/engine_state/query.rs
	execution_engine/src/core/engine_state/run_genesis_request.rs
	execution_engine/src/core/engine_state/step.rs
	execution_engine/src/core/engine_state/upgrade.rs
	execution_engine/src/core/execution/address_generator.rs
	node/src/components/rpc_server/rpcs/state.rs
	smart_contracts/contract/src/contract_api/mod.rs
	types/src/cl_type.rs
	types/src/contracts.rs

Commit:
Fix JSON fixtures after changing schema.

luis-casper committed:
Documenting undefined behavior.

dwerner commented on execution_engine/src/core/engine_state/execution_result.rs:
```diff
@@ -315,6 +315,9 @@ impl ExecutionResult {
     /// account's main purse, and add `max_payment_cost` to proposer account's balance.
+    ///
+    /// NOTE: This function assumes that `max_payment_cost <= account_main_purse_balance`.
+    /// Not obeying this restriction is a programming error with undefined behavior.
```
While it is currently possible to underflow as indicated by the comment, wouldn't it be worthwhile to implement `Motes::checked_sub()` and use that instead of the bare subtraction in `make_payment_error_effects`? @Fraser999 @mpapierski

mpapierski replied:
@dwerner @Fraser999 addressed in ac7706d77c601fd75d8f09030f64369aec2a5a23 and ticketed bigger scope of removal of unchecked ops in https://github.com/casper-network/casper-node/issues/2103.

Commit:
Fix one instance of unchecked arithmetic.

dwerner commented on execution_engine/src/core/engine_state/execution_result.rs:
```diff
@@ -17,13 +17,15 @@ fn make_payment_error_effects(
-    let new_balance = account_main_purse_balance - max_payment_cost;
+    let new_balance = account_main_purse_balance
+        .checked_sub(max_payment_cost)
+        .ok_or(error::Error::InsufficientPayment)?;
```
:+1:

Fraser999 commented on execution_engine_testing/test_support/src/lib.rs:
```diff
@@ -64,7 +64,6 @@
     html_logo_url = "https://raw.githubusercontent.com/CasperLabs/casper-node/master/images/CasperLabs_Logo_Symbol_RGB.png",
     test(attr(forbid(warnings)))
 )]
-#![warn(missing_docs)]
```
Removed by mistake?

Fraser999 commented on node_macros/src/lib.rs:
```diff
@@ -6,7 +6,7 @@
     test(attr(forbid(warnings)))
 )]
-#![warn(missing_docs, trivial_casts, trivial_numeric_casts)]
+#![warn(trivial_casts, trivial_numeric_casts)]
```
Removed by mistake?

Fraser999 commented on smart_contracts/contract/src/lib.rs:
```diff
@@ -56,7 +56,6 @@
     html_logo_url = "https://raw.githubusercontent.com/CasperLabs/casper-node/master/images/CasperLabs_Logo_Symbol_RGB.png",
     test(attr(forbid(warnings)))
 )]
-#![warn(missing_docs)]
```
Removed by mistake?

mpapierski replied:
Addressed in 547cd7969e4c2e1af93218cece1de0be75966026

Fraser999 reviewed (approved):
Three lints removed by mistake by the looks of it.  Otherwise this looks great to me.

Approved assuming the lints are reinstated.

Commit:
Fix missing directives

mpapierski commented:
bors r+
